zebra-consensus/src/block/subsidies.rs

```diff
@@ -0,0 +1,236 @@
+//! Calculation of Block Subsidy, Funding Streams, and Founders’ Reward
+
+use zebra_chain::{
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::parameters::{fs, Params};
+
+fn slow_start_shift() -> Height {
+    Height(Params::SLOW_START_INTERVAL / 2)
+}
+
+fn slow_start_rate() -> u32 {
+    Params::MAX_BLOCK_SUBSIDY / Params::SLOW_START_INTERVAL
+}
+
+fn halving(height: Height, network: Network) -> u32 {
+    if height < slow_start_shift() {
+        return 0;
+    }
+    Blossom
+        .activation_height(network)
+        .filter(|&blossom_height| height >= blossom_height)
+        .map(|blossom_height| {
+            (blossom_height - slow_start_shift()) as u32 * Params::BLOSSOM_POW_TARGET_SPACING_RATIO
+                + height.0
+                - blossom_height.0
+        })
+        .map(|scaled_halvings| scaled_halvings / Params::POST_BLOSSOM_HALVING_INTERVAL)
+        .unwrap_or_else(|| {
+            (height - slow_start_shift()) as u32 / Params::PRE_BLOSSOM_HALVING_INTERVAL
+        })
+}
+
+pub fn block_subsidy(height: Height, network: Network) -> u32 {
+    if height < slow_start_shift() {
+        slow_start_rate() * height.0
+    } else if slow_start_shift() <= height && height < Height(Params::SLOW_START_INTERVAL) {
+        slow_start_rate() * (height.0 + 1)
+    } else {
+        Blossom
+            .activation_height(network)
+            .filter(|&blossom_height| {
+                height < blossom_height && Params::SLOW_START_INTERVAL <= height.0
+            })
+            .map(|_| Params::MAX_BLOCK_SUBSIDY >> halving(height, network))
+            .unwrap_or_else(|| {
+                ((Params::MAX_BLOCK_SUBSIDY / Params::BLOSSOM_POW_TARGET_SPACING_RATIO
+                    * 2u32.pow(halving(height, network))) as f32)
+                    .floor() as u32
```

Comment:
Please don't use floating point to calculate exact amounts of money. Floating point values can have surprising rounding and precision.

Instead, carefully use integer multiplication and division to get the same results as the specification, and write a comment explaining how they match the spec. Remember that integer division truncates - it does an automatic floor.

```suggestion
                    * (1 << halving(height, network))))
```